src/app/learnocaml_exercise_main.ml

```diff
@@ -93,6 +93,172 @@ let display_report exo report =
     (Format.asprintf "%a" Report.(output_html ~bare: true) report) ;
   grade
 
+let display_descr ex_meta =
+  let open Tyxml_js.Html5 in
+  let open Learnocaml_data.Exercise in
+  div ~a:[ a_class [ "descr" ] ] [
+    h2 ~a:[ a_class [ "learnocaml-exo-meta-category" ] ] [ pcdata ex_meta.Meta.title ] ;
+    p [ match ex_meta.Meta.short_description with
+        | None -> pcdata [%i"No description available."]
+        | Some text -> pcdata text ] ;
+  ]
+
+let display_stars ex_meta =
+  let open Tyxml_js.Html5 in
+  let open Learnocaml_data.Exercise in
+  div ~a:[ a_class [ "stars" ] ] [
+    let num = 5 * int_of_float (ex_meta.Meta.stars *. 2.) in
+    let num = max (min num 40) 0 in
+    let alt = Format.asprintf [%if"difficulty: %d / 40"] num in
+    let src = Format.asprintf "icons/stars_%02d.svg" num in
+    img ~alt ~src ()
+  ]
+
+let display_kind ex_meta =
+  let open Tyxml_js.Html5 in
+  let open Learnocaml_data.Exercise in
+  div ~a:[ a_class [ "length" ] ] [
+    match ex_meta.Meta.kind with
+    | Exercise.Meta.Project -> pcdata [%i"project"]
+    | Exercise.Meta.Problem -> pcdata [%i"problem"]
+    | Exercise.Meta.Exercise -> pcdata [%i"exercise"] ]
+
+(* Inefficient version: Reload the exercise description each time. *)
+let display_exercise_more content_id token id =
+  let open Tyxml_js.Html5 in
+  let open Learnocaml_data.Exercise in
+  let content = find_component content_id in
+  token >>= fun token ->
+  Server_caller.fetch_exercise token id >>= fun (meta, _, _) ->
+  let descr =
+    a ~a:[ a_href ("exercise.html#id=" ^ id ^ "&action=open") ;
+           a_class [ "exercise" ] ] [
+      display_descr meta ;
+      div ~a:[  ] [
+        display_stars meta ;
+        display_kind meta ;
+      ] ]
+  in
+  Manip.replaceChildren content [ descr ];
+  Lwt.return ()
+
+let display_list ?(sep=Tyxml_js.Html5.pcdata ", ") l =
+  let open Tyxml_js.Html5 in
+  let rec gen acc = function
+    | [] -> [ pcdata "" ]
+    | a :: [] -> a :: acc
+    | a :: ((_ :: _) as rem) ->
+        gen (sep :: (a  :: acc)) rem
+  in
+  gen [] l |> List.rev
+
+
+let display_authors authors =
+  let open Tyxml_js.Html5 in
+  let author (name, mail) =
+    span [ pcdata name;
+           pcdata " <";
+           a ~a:[ a_href ("mailto:" ^ mail) ]
+             [ pcdata mail ];
+           pcdata ">"
+         ] in
+  display_list @@ List.map author authors
+
+let display_skill_more skill content_id =
+  let open Tyxml_js.Html5 in
+  let open Learnocaml_data.Exercise in
+  let content = find_component content_id in
+  let req = match skill with
+      `Focus s -> Learnocaml_api.Focusing_skill s
+    | `Requirements s -> Learnocaml_api.Requiring_skill s
+  in
+  Server_caller.request_exn req >>= fun exs ->
+  Manip.replaceChildren content @@ display_list @@ List.map pcdata exs;
+  Lwt.return ()
+
+let display_more token ex_meta id =
+  let open Learnocaml_data.Exercise in
+  let open Tyxml_js.Html5 in
+  let display_skill_link content_id s =
```

Comment:
The next two functions seem to share a common structure. Can't you factorize them?